### PR TITLE
Extensions for the AppProviders to improve the UI

### PR DESCRIPTION
**Pull request by glpatcern**
This PR introduces further extensions for the open-in-app workflow:
* `AppProvider`s expose a `capability` to identify fully fledged editors vs viewers
* Mime types are defined including a friendly name and a corresponding file extension

This allows the UI to e.g. populate a **New** menu with all types for which an `editor` AppProvider is defined.

The extensions constitute breaking changes for the protobuf bindings, and can be merged once cs3org/reva#2067 is ready to be merged.

### CODE REVIEW

**Review by labkode — APPROVED** [review]
Looks good. In general is better to not break backwards compatibility by maintaining the order of the fields in the protobuf definitions. However as this PR introduces a new REQUIRED fields it does not change anything.

**Comment by glpatcern** [comment]
Yeah I thought about that and also about leaving them OPTIONAL but in the end we need them...

File: cs3/app/registry/v1beta1/resources.proto
Comment (wkloucek) [inline cs3/app/registry/v1beta1/resources.proto:52]: How exactely would the UI make use of the capabilities? 
If it has `CAPABILITY_VIEWER` obviously shouldn't offer to edit a file... But if something is a `CAPABILITY_EDITOR` a user might also want to know if it's possible to open a file in reading mode.
Shouldn't the Capabilities reflect all possible OpenInAppRequest.ViewMode(s) https://cs3org.github.io/cs3apis/#cs3.gateway.v1beta1.OpenInAppRequest.ViewMode?

Comment (glpatcern) [reply cs3/app/registry/v1beta1/resources.proto:52]: This was discussed in a now merged PR. There are two motivations for this:
* In the context menu for files, one could show `Preview in <app>` vs both `Open in <app>` and `Preview`. Of course if the user has no write access, only `Preview` would be shown.
* Populate the "New" button menu with mime types for which an AppProvider is registered with `EDITOR` capability. As a (counter-)example, assuming that e.g. Collabora is able to open jpeg files in view mode, do not show it in the New list.

Now the first is not super urgent in the current state of integrations, whereas the second is quite critical. The other part of this PR - providing a list of "well-known" mime types with their description and corresponding file extensions - specifically targets that.

Finally, I did think about some "beauty of symmetry" to make `capabilities` reflect the `viewMode` values, but I don't see the added value - for the UI you just need to know if an app is capable of editing or not, a viewer with or without a download button (and with or without the ability to hide it should the permission be so restrictive) is still a viewer and would be fired from a `Preview` menu item.

Comment (glpatcern) [reply cs3/app/registry/v1beta1/resources.proto:52]: Another upfront consideration is the fact that we do have a number of "UI-only" apps (e.g. drawio, the pdfviewer, etc.) which are not captured by the AppProviders scheme. So anyway the UI must take those into account differently.

**Comment by glpatcern** [comment]
Self-reviewing: this model is still missing the ability to register the mime type attributes (name, ext, icon...) from "outside" (e.g. the configuration). A `SetMimeType` RPC method may be needed for this.

Also, I don't understand why such a large diff in `proto.lock`. I dropped and recreated it as this is a breaking change, does that mean that we were carrying over lots of "garbage"?

File: Makefile
Comment (labkode) [inline Makefile:11]: Explanation please :)

File: cs3/app/provider/v1beta1/provider_api.proto
Comment (labkode) [inline cs3/app/provider/v1beta1/provider_api.proto:74]: You can add an optional field named "template". Looking forward in case we want to provide custom templates (like CERN branded Collabora files).

File: Makefile
Comment (glpatcern) [reply Makefile:11]: :-D sure - I'll add a comment in the file

File: cs3/app/provider/v1beta1/provider_api.proto
Comment (labkode) [inline cs3/app/provider/v1beta1/provider_api.proto:58]: From the UI perspective. As a I user I should be able to create for example: "New MS Word document" or "New OnlyOffice document". I think we may need to add the `app` name in the payload.

File: cs3/app/provider/v1beta1/provider_api.proto
Comment (glpatcern) [reply cs3/app/provider/v1beta1/provider_api.proto:74]: Makes sense, as eventually the user chooses between e.g. `New Word Document` or `New Word Document from corporate template`...

File: cs3/app/provider/v1beta1/provider_api.proto
Comment (glpatcern) [reply cs3/app/provider/v1beta1/provider_api.proto:58]: Just discussed with Lisa and Diogo: the creation of the file (either touch or copy from template) takes place before firing the app, and it does not involve the app at all - at least, so far we don't know any app (definitely not the WOPI ones) that expose an endpoint to _create_ a new file. The new file/template operation is therefore only on the storage and app independent.

**Comment by glpatcern** [comment]
Will merge after https://github.com/cs3org/cs3apis/pull/145 works

File: cs3/app/registry/v1beta1/resources.proto
Comment (wkloucek) [reply cs3/app/registry/v1beta1/resources.proto:52]: > Another upfront consideration is the fact that we do have a number of "UI-only" apps (e.g. drawio, the pdfviewer, etc.) which are not captured by the AppProviders scheme. So anyway the UI must take those into account differently.

That's true but these apps are also ownCloud Web specific and can't be used in any other web frontend. The app driver returns generic application which can also be used in a different web frontend. Therefore it is ok to just list what the backend / reva provides.

File: cs3/app/registry/v1beta1/registry_api.proto
Comment (ishank011) [inline cs3/app/registry/v1beta1/registry_api.proto:140]: Instead of `MimeTypeList`, we can just use `repeated MimeTypeInfo` here. What do you think?

**Comment by ishank011** [comment]
Also, you'll need to add the proto lock file for the build to pass

File: cs3/app/registry/v1beta1/registry_api.proto
Comment (glpatcern) [reply cs3/app/registry/v1beta1/registry_api.proto:140]: It's equivalent indeed, but isn't it convenient to have a type for the list in Reva? or in that case we just have a `map`?

Comment (ishank011) [reply cs3/app/registry/v1beta1/registry_api.proto:140]: Yep, we had to use this with a map because `map<string, repeated message>` isn't supported. It adds one more layer of abstraction which doesn't have much use. The access now would look like `response.MimeTypes.MimeTypes`, when we can have `response.MimeTypes`.

**Pull request merged**
